### ISSUE TITLE
Change the look of voted answer

### ISSUE THREAD

**Issue by kaakaa** [matterpoll/matterpoll#28]
We cannot see on browser which answers I voted.
So the look of voted answer should be changed. (e.g: Change background color, Add (*) as prefix, ...)

Sample image: 
![2018-08-24 13 22 52](https://user-images.githubusercontent.com/1453749/44565126-894e8d80-a7a1-11e8-99e2-e27c92571169.png)

This feature may need to implement webapp plugin, because UI changes are different for each user.

**Comment by Hanzei** [comment]
I like the idea but I'm not sure if its possible to create this. Its not possible to change the text or color for the server side. It might work with webapp but I'm sceptical.

**Comment by kaakaa** [comment]
Yes, I don't know how can we implement this now. But I think this feature is necessary.
I'll investigate more about webapp plugin.

**Comment by hanzei** [comment]
Since https://github.com/mattermost/mattermost-server/pull/9859 got merged: Are there more changes in Mattermost required?

**Comment by kaakaa** [comment]
Thank you for letting me know it!
I don't think that more changes in Mattermost is needed.

And now, [my super early prototype](https://github.com/matterpoll/matterpoll/commit/08bc956b27d77fce8beb68c4bc0d97311d3d1662) can change style of voted answer.
The followings are things to consider.

### Architecture
For achiving this feature, I'll add `answers` props to the post and parse it on webapp plugin.

```
"props": {
    "attachments": {...},
    "answers": {
        "answer_option1": ["user_id1"],
        "answer_option2": ["user_id2", "user_id3"],
        ...
    }
}
```

### Anonymous Settings
Since all clients receive `answers` props, anonymous setting has less meanings.
Of course, `answers` info isn't displayed, but I think that users can get `answers` info with modifying js.

### AnswerOptionId
Now, the voted answer is distinguished by answer string. So if answer strings are duplicated, Matterpoll can't change the style of buttons properly.
We should add OptionId to AnswerOption or reject poll that has duplicated answer string.



How do these sounds for you?

**Comment by hanzei** [comment]
First of all the prototype is awesome!  :tada: 

#### Architecture

We could also add a new API endpoint and fetch this information via http. This might be a cleaner approach. 

#### Anonymous Settings

This might be a concern. The APi approach would help here.

> Now, the voted answer is distinguished by answer string. So if answer strings are duplicated, Matterpoll can't change the style of buttons properly.
We should add OptionId to AnswerOption or reject poll that has duplicated answer string.

I think we should reject duplicate anwer strings. See #115.

**Comment by hanzei** [comment]
Your approach is fine. We could do something more generic like GET `/polls/{id:[a-z0-9]+}/votes`. But 0/5.

**Comment by kaakaa** [comment]
I start implementing new API for fetching voted answers.
For implementing new API, webapp needs to have `poll_id`. 
It may lead concerns about security (described in #61). If `poll_id` is exposed, polls become controllable to users who can get `user_id` of poll creator or admin.

As a possible idea, we may be able to create new API with `action_id`, no `poll_id`. But APIs related to `action_id` are not public API, so this approach doesn't look good.

@hanzei How do you think of this?

**Comment by hanzei** [comment]
I knew this would be a problem at some point...

With the current system only the Mattermost Server and Matterpoll know a polls id. This changes with the introduction of webapp components because the webapp need to know a poll id and therefore the user can extract this id.

I don't thing adding a new APi helps here. There will be a point where the webapp part want to do API request that change data.

Two ideas:
- We implement the secure mechanisms proposed in #61.
- We check for HTTP header `Mattermost-User-Id` in the API. It is only present when a Mattermost user is logged in. This way we prevent attacks from outside. Of course this won't help for open servers.

**Comment by kaakaa** [comment]
Thanks for your feedbacks.

Hmm...I think that first idea doesn't make security strong. If attackers get poll_id, they can also get secure token in the same way.

Second idea makes sense. 
`Mattermost-User-Id` users set is removed [here](https://github.com/mattermost/mattermost-server/blob/master/app/plugin_requests.go#L65) before plugin receives requests. So attackers must get access token for impersonating by `Mattermost-User-Id` header. It looks secure.

> Of course this won't help for open servers.

But I can't imagine what this line warns. Could you give me more words?

**Comment by hanzei** [comment]
> > Of course this won't help for open servers.
> 
> But I can't imagine what this line warns. Could you give me more words?

I'm talking about server where [`EnableOpenServer`](https://docs.mattermost.com/administration/config-settings.html#enable-open-server) is set to `true` like https://community.mattermost.com.

**Comment by kaakaa** [comment]
Thank you for comments and sorry for my late reply.
I see. The approach using `Mattermost-User-Id` seems not to be perfect.

> > * We implement the secure mechanisms proposed in #61.

> Hmm...I think that first idea doesn't make security strong. If attackers get poll_id, they can also get secure token in the same way.

But this my opinion seems to be incorrect. 
If setting secure token in [PostActionIntegration.Context](https://github.com/mattermost/mattermost-server/blob/master/model/integration_action.go#L44), attacker can't know secure token. And if polls are updated only when poll_id and secure token are correct, I think security problems will not occur.
So we can disclose poll_id to client and fetch voted answers via new API.

Is this right? If so, I'll work to be done #61 at first.

**Comment by hanzei** [comment]
> Thank you for comments and sorry for my late reply.
I see. The approach using `Mattermost-User-Id` seems not to be perfect.

I was wrong on this `Mattermost-User-Id` will always be set the the user who's making the request. We just have to check if `Mattermost-User-Id` is the same as the creator id and we are fine.

For "public" API endpoints like the above proposed `/polls/{id:[a-z0-9]+}/votes` we can just verify that `Mattermost-User-Id` is set. This way we allow every logged in user to see every votes from every poll with a known plugin id. But I think this is fine for the moment. We can narrow the scope later with something like [`HasPermissionToChannel()`](https://developers.mattermost.com/extend/plugins/server/reference/#API.HasPermissionToChannel).

**Comment by kaakaa** [comment]
> I was wrong on this. Mattermost-User-Id will always be set the the user who's making the request. We just have to check if Mattermost-User-Id is the same as the creator id and we are fine.

I found that requests via PostAction don't have `Mattermost-User-Id` header, even if requests are sent to endpoints of plugnis (sorry, I missed this spec). Session info is dropped in [`app/integration_action.go`](https://github.com/mattermost/mattermost-server/blob/master/app/integration_action.go#L128)
So we can't verify requests from PollOption buttons by user_id.

We could choose one from the following.
* Introduce secure token (this will break all existing polls)
* Introduce new pseudo ID to inquire about voted items to each vote

**Comment by hanzei** [comment]
> > I was wrong on this. Mattermost-User-Id will always be set the the user who's making the request. We just have to check if Mattermost-User-Id is the same as the creator id and we are fine.
> 
> I found that requests via PostAction don't have `Mattermost-User-Id` header, even if requests are sent to endpoints of plugnis (sorry, I missed this spec). Session info is dropped in [`app/integration_action.go`](https://github.com/mattermost/mattermost-server/blob/master/app/integration_action.go#L128)

> So we can't verify requests from PollOption buttons by user_id.

That is not true. We can get the userID from the `PostActionIntegrationRequest` or the `PostActionIntegrationRequest`. We only need the `Mattermost-User-Id` header for http requests made by the webapp directly.

**Comment by kaakaa** [comment]
Yes,  I know that we can get userID from `PostActionIntegrationRequest`. But unlike `Mattermost-User-Id`, userId in `PostActionIntegrationRequest` can be set easily by request body.  (e.g.: `curl https://example.com/plugins/matterpoll/api/v1/polls/${POLL_ID}/end -d '{"user_id": "${CREATE_ID}"}'`)

So if exposing poll_id, anyone who knows pollId and userId can impersonate others.
Do we allow this for now?

**Comment by hanzei** [comment]
You right, action requests are currently only secured by the `userID` in the body. But we could use the "static secure token" approach here. And the Mattermost team might accept a PR which adds `Mattermost-User-Id` to requests made by `DoActionRequest()`.

To sum to the discussion: Declaring pollIDs as a public available information weakens the security of action button requests, but we can work on this with the above proposed mechanisms. API requests made by the webapp part can be secured by checking `Mattermost-User-Id`.

@kaakaa Do you agree with this summary?

**Comment by kaakaa** [comment]
Thanks a lot! You summary made my understanding clear.
It means that we start making PR which adds `Mattermost-User-Id` to requests made by PostAction, doesn't you?  And after merging PR, we change how to get user_id from `PostActionINtegrationRequest` to `Mattermost-User-Id` header, and only logged in user can control poll.

If my understanding is well, I'll work for making PR 👍

**Comment by hanzei** [comment]
Thanks @kaakaa! It might be worth to discuss this in the Developers Toolkit channel.

**Comment by kaakaa** [comment]
Closed by #160
